nft_ingester/src/bin/slot_persister/main.rs

```diff
@@ -338,71 +346,148 @@ async fn process_slots(
+                                new_failed_slots.extend_from_slice(&slots_to_fetch);
+                            },
+                        }
+                    }
+                    new_failed_slots.sort();
+                    new_failed_slots.dedup();
+                },
+
+                // ---------------------------------------------------------
+                // 2) RPC or other: original slot-by-slot concurrency
+                // ---------------------------------------------------------
+                _ => {
+                    let fetch_futures = slots_to_fetch.iter().map(|&slot| {
+                        let backfill_source = backfill_source.clone();
+                        let semaphore = semaphore.clone();
+                        let shutdown_token = shutdown_token.clone();
+
+                        async move {
+                            let _permit = semaphore.acquire().await;
+                            fetch_block_with_retries(backfill_source, slot, shutdown_token).await
+                        }
+                    });
+
+                    let results = join_all(fetch_futures).await;
+                    for result in results {
+                        match result {
+                            Ok((slot, raw_block)) => {
+                                successful_blocks.insert(slot, raw_block);
+                            },
+                            Err((slot, e)) => {
+                                error!("Failed to fetch slot {}: {:?}", slot, e);
+                                new_failed_slots.push(slot);
+                            },
+                        }
+                    }
+                },
             }
 
             if new_failed_slots.is_empty() {
-                // All slots fetched successfully, save to database
                 debug!(
-                    "All slots fetched successfully for current batch. Saving {} slots to RocksDB.",
+                    "All slots fetched in this batch. Attempting to save {} blocks to RocksDB...",
                     successful_blocks.len()
                 );
-                if let Err(e) = target_db.raw_blocks_cbor.put_batch(successful_blocks.clone()).await
-                {
-                    error!("Failed to save blocks to RocksDB: {}", e);
-                    // Handle error or retry saving as needed
-                    batch_retries += 1;
-                    if batch_retries >= MAX_BATCH_RETRIES {
-                        panic!(
-                            "Failed to save batch to RocksDB after {} retries. Discarding batch.",
-                            MAX_BATCH_RETRIES
-                        );
-                    } else {
-                        warn!(
-                            "Retrying batch save {}/{} after {} ms due to error.",
-                            batch_retries, MAX_BATCH_RETRIES, batch_delay_ms
+
+                let projected_last_slot = successful_blocks.keys().max().copied().unwrap_or(0);
```

Comment:
Is not `copied()` redundant?

Reply:
the `keys().max()` thing returns a reference, and to do `unwrap_or(0)` we need to get an owned value

Reply:
We do not. `successful_blocks.keys().max().unwrap_or(&0);`. However, `successful_blocks` needs to be mutable, so copying is fine.